Cubemap.cpp

```diff
@@ -411,6 +414,427 @@ Cubemap* Cubemap::shFilterCubeMap(bool useSolidAngleWeighting, int fixup, int ou
+        width = mipMapWidth[ maxRealMipStart - m ];
+        lastPixelOffset = ((y * 2 * width) + x * 2) * stride;
+
+        maxLum = -1;
+        maxLumNum = 0;
+        // check the Four Neighbor Only
+        // because we come from mip above
+        // which is box averaged from those 4
+        for ( k = 0; k < 2; k++ )
+        {
+            for ( p = 0; p < 2; p++ )
+            {
+                // each Mipmap, check brightest
+                const uint offset = lastPixelOffset + p * stride + k * stride * width;
+                pixel = luminance( pixels[ offset ], pixels[ offset + 1 ], pixels[ offset + 2 ] );
+
+                if ( pixel == maxLum ) {
+
+                    maxLumNum++;
+
+                } else if ( pixel > maxLum ) {
+
+                    maxLumNum = 1;
+                    maxLum = pixel;
+                    // new start where to search
+                    // on bigger mipmap
+                    maxPixelOffset = offset;
+
+                }
+            }
+        }
+
+
+        // store value at this mip
+        mipMapMaxLum[ maxRealMipStart - m ] = maxLum;
+        mipMapLumNum[ maxRealMipStart - m ] = maxLumNum;
+        mipMapMaxPixel[ maxRealMipStart - m ] = maxPixelOffset;
+
+        if ( maxLumNum > 1 ) {
+
+            // too many pixels at same Max Luminosity
+            // we revert to inferior mipmap
+            // to get a mipmap where lum max is 1 pixel
+            //(means it's area light of same pixel values)
+            break;
+
+        }
+
+        // store as last acceptable results
+        // if next one has too many bright pixel equalities
+        // ( that's why after the break above)
+        mipMapLum = maxRealMipStart - m;
+    }
+
+    // we have a winner here
+    maxPixelOffset = mipMapMaxPixel[ mipMapLum ];
+    maxLum = mipMapMaxLum[ mipMapLum ];
+    width = mipMapWidth[  mipMapLum ];
+
+    // getDirection from pixel position
+    // The greatest magnitude component, S, T or R,
+    // select the cube face.
+    // The other 2 components are used to get texel
+    pixelPos = maxPixelOffset / stride;
+    x = pixelPos % width;
+    y = floor( pixelPos / width );
+
+    Vec3f direction;
+    Vec3f colorAvg;
+
+    texelCoordToVect(maxFace, (float)x, (float)y, width, &direction[0], 1);
+
+    //Store Light info
+    // Lat, Long ? no use now but...
+    std::cout << "{ ";
+    std::cout << "  x: "  << x;
+    std::cout << ", y: "  << y;
+    std::cout << ", z: " << maxFace;
+    std::cout << ", face: " << maxFace;
+    std::cout << ", mipMapOffset: " << mipMapLum;
+    std::cout << ", maximumPixelOffset: " << maxPixelOffset;
+    std::cout << ", maximumLuminosity: " << maxLum;
+    std::cout << ", color: [" << colorAvg[0] << ", " << colorAvg[1] << ", " << colorAvg[2] << "]";
+    std::cout << ", direction: [" << direction[0] << ", " << direction[1] << ", " << direction[2] << "]";
+    std::cout << " } ";
+
+    std::cout << std::endl;
+
+#if !defined(NDEBUG)
```

Comment:
only save the debug texture in debug mode, just to be sure it doesn't spill debug texture in release mode